src/content/actions/receive-profile.js

```diff
@@ -364,51 +401,64 @@ export function receiveProfileFromFile(profile: Profile): ThunkAction {
   };
 }
 
-export function errorReceivingProfileFromFile(error: any): Action {
+export function errorReceivingProfileFromFile(error: Error): Action {
   return {
     type: 'ERROR_RECEIVING_PROFILE_FROM_FILE',
     error,
   };
 }
 
+function _fileReader(input) {
+  const reader = new FileReader();
+  const promise = new Promise((resolve, reject) => {
+    reader.onload = () => resolve(reader.result);
+    reader.onerror = () => reject(reader.error);
+  });
+
+  return {
+    asText() {
+      reader.readAsText(input);
+      return promise;
+    },
+
+    asArrayBuffer() {
+      reader.readAsArrayBuffer(input);
+      return promise;
+    },
+  };
+}
+
 export function retrieveProfileFromFile(file: File): ThunkAction {
-  return dispatch => {
+  return async dispatch => {
     dispatch(waitingForProfileFromFile());
 
-    (new Promise((resolve, reject) => {
-      const reader = new FileReader();
-      reader.onload = () => resolve(reader.result);
-      reader.onerror = reject;
-      reader.readAsText(file);
-    })).then(text => {
+    try {
+      const text = await _fileReader(file).asText();
       const profile = unserializeProfileOfArbitraryFormat(text);
       if (profile === undefined) {
         throw new Error('Unable to parse the profile.');
       }
 
       dispatch(receiveProfileFromFile(profile));
-    }).catch(() => {
-      return (new Promise((resolve, reject) => {
-        const reader = new FileReader();
-        reader.onload = () => resolve(reader.result);
-        reader.onerror = reject;
-        reader.readAsArrayBuffer(file);
-      }));
```

Comment:
note: there was in issue in the previous code here: we should have returned here instead of continuing in the promise chain. The new code fixes this.